### DISCUSSION TITLE
Making request with urequests

### DISCUSSION THREAD

**Discussion by sharhan-alhassan** [micropython/micropython#11205] [Using MicroPython]
Hey Team,
I have configured a Kubernetes kind cluster to run a flask API. The flask App when run locally with flask run produces an IP which is the local machine assigned IP from the local network (192.168.100.19). My ESP8266 which also has an IP of (192.168.100.36) can talk to the Flask API.

However, things are different when I run the Flask app from the kind cluster. The kind nodes are in the network of 192.168.64.1 so my nodes' IPs are in that range. I used metallb to create the IP pools in that range as well so my Flask API is given a Loadbalancer IP of 192.168.63.200 This Flask API IP cannot be communicated with the Nodemcu IP of 192.168.100.36

The Genesis of the issue is that, the ESP8266 cannot even make a GET request to any public API like getting cat facts from "https://catfact.ninja/fact" It can only communicate with a local running flask API

Below is my code logic to send a post request to a Flask API

```py
import time
import ntptime
import network
import json
import utime
import urequests as requests
from machine import Pin

# Working connection string for flask API running locally
CONNECTION_STRING = 'http://192.168.100.19:5000/api/post'
# ESP8266 IP from ifconfig is 192.168.100.38

# Define the pin numbers for the ultrasonic sensor
TRIG_PIN = 12
ECHO_PIN = 14

wlan = network.WLAN(network.STA_IF)
wlan.active(True)
if not wlan.isconnected():
    print('connecting to network...')
    wlan.connect('ssid', 'password')
    while not wlan.isconnected():
        pass
    print('network config:', wlan.ifconfig())

# Define the message sender function
def send_message(distance):
    try:
        ntptime.settime()
        # Create a JSON payload with the sensor data
        localtime = utime.localtime()
        timestamp = '{}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}Z'.format(
            localtime[0], localtime[1], localtime[2],
            localtime[3], localtime[4], localtime[5])
        payload = {
            "distance": distance,
            "timestamp": timestamp
        }

        # Send the payload to the Flask app
        response = requests.post(
            CONNECTION_STRING, 
            headers = {
            'content-Type': 'application/json',
            },
            data=json.dumps(payload)
        )

        # Print the response status code
        print('Response:', response.status_code)

    except Exception as e:
        print('Error sending message:', e)

# Set up the ultrasonic sensor pins
trig = Pin(TRIG_PIN, Pin.OUT)
echo = Pin(ECHO_PIN, Pin.IN)

# Loop indefinitely to send sensor data every 5 seconds
while True:
    # Trigger ultrasonic sensor to measure distance
    trig.value(0)
    time.sleep_us(2)
    trig.value(1)
    time.sleep_us(10)
    trig.value(0)

    # Calculate distance from ultrasonic sensor data
    pulse_start = time.ticks_us()
    while echo.value() == 0:
        pulse_start = time.ticks_us()
    pulse_end = time.ticks_us()
    while echo.value() == 1:
        pulse_end = time.ticks_us()
    pulse_duration = pulse_end - pulse_start
    distance = pulse_duration / 58.0

    # Send sensor data to IoT Hub
    send_message(distance)
    time.sleep(4)

```

**Comment by sharhan-alhassan** [comment]
@anotherjesse @kumekay @kadamski @esmil

**Comment by karfas** [comment]
> - My ESP8266 which also has an IP of (192.168.100.36) can talk to the Flask API.

So your app is working fine, doesn't look like a micropython issue.

> - Loadbalancer IP of 192.168.63.200 This Flask API IP cannot be communicated with the Nodemcu IP of 192.168.100.36
> - cannot even make a GET request to any public API like getting cat facts from "https://catfact.ninja/fact"

What exactly do you get from the `print('Error sending message:', e)` statement ?

This smells like a routing and/or firewall problem (but I have not looked if/how urequests handles https:// URLs).

Can you send requests using the browser on your mobile phone (or a laptop) connected to the same access point ? 

Maybe your cluster disallows requests from outside it's own network ?

**Comment by sharhan-alhassan** [comment]
Hey @karfas Thanks for the response.
`What exactly do you get from the print('Error sending message:', e) statement ?`
This is the error I'm getting when I'm trying to connect to the service IP (created by mettalb)
```sh
Error sending message: [Errno 103] ECONNABORTED
Error sending message: [Errno 103] ECONNABORTED
```

On the other hand, like I said, I tried this simple GET request to the cat facts API and this is this error from the ESP8266 device

* Simple GET request just to test if the device can communicate to any URL aside an endopoint in the same network
```py
url = 'https://catfact.ninja/fact'

response = requests.get(url)

if response.status_code == 200:
    json_data = json.loads(response.content)
    print(json_data)
else:
    print('Failed to get data. Status code:', response.status_code)

```

Error I'm getting
```sh
File "<stdin>", line 1, in <module>
  File "urequests.py", line 116, in get
  File "urequests.py", line 62, in request
OSError: -40
```

**Reply by karfas** [reply]
**And what do you find in urequests.py, in lines 62 and 116 ?**

The current version of urequest - found in https://github.com/micropython/micropython-lib/blob/master/micropython/urllib.urequest/urllib/urequest.py - has no line 116 and shows me at line 62 
```py
               raise NotImplementedError("Redirects not yet supported")
```
A https request is not "simple", as it requires some additional memory and the TLS handshake with the server.

**Reply by karfas** [reply]
`ECONNABORTED` is caused most likely somewhere on the server side in the `accept()` call.

Have you already checked your network configuration ?

**Reply by sharhan-alhassan** [reply]
Hey @karfas 
* I can't actually see any `urequests.py` from my end. I'm using Arduino Lab for Python ( on Ubuntu) -- only seeing `boot.py`
* When I use `ngrok` utility to forward the API and supply the URL to the ESP8266, it works actually
